python/lsst/ts/eas/diurnal_timer.py

```diff
@@ -188,85 +187,153 @@ def __init__(self, sun_altitude: str | float = -18.0):
 
         self.sun_altitude = sun_altitude
         self.twilight_time: Time | None = None
+        self.sunrise_time: Time | None = None
 
         self._noon_loop_ready = asyncio.Event()
         self._twilight_loop_ready = asyncio.Event()
+        self._sunrise_loop_ready = asyncio.Event()
 
         self.noon_condition = asyncio.Condition()
         self.twilight_condition = asyncio.Condition()
+        self.sunrise_condition = asyncio.Condition()
 
         self.is_running = False
         self._tasks: list[asyncio.Task] = []
 
-    async def _run_noon_loop(self) -> None:
-        """Notifies noon_condition whenever it's noon.
+    @staticmethod
+    def _seconds_until(target: Time, now: Time | None = None) -> float:
+        """Returns time from `now` until `target` in seconds.
 
-        This method loops, notifying for noon_condition each day.
-        """
-        while self.is_running:
-            local_noon = get_local_noon_time()
-            wait_seconds = (local_noon - Time.now()).sec
-            self._noon_loop_ready.set()
+        If the target has already passed, zero is returned to avoid
+        passing a negative value to `sleep`.
 
-            await asyncio.sleep(wait_seconds)
+        Parameters
+        ----------
+        target: `~astropy.time.Time`
+            The future event that is to be timed.
 
-            async with self.noon_condition:
-                self.noon_condition.notify_all()
+        now: `~astropy.time.Time`
+            The starting time, or None if `Time.now()` should be used.
 
-    def get_twilight_time(self, after: Time) -> Time:
-        """Returns the time of next end of twilight for the given time.
+        Returns
+        -------
+        float
+            Number of seconds from `now` until `target`.
+        """
+        now = now or Time.now()
+        dt = (target - now).sec
+        return max(60.0, dt)
 
-        Parameters
-        ----------
-        after : `~astropy.time.Time`
-            Time returned will be time corresponding to the
-            end of twilight on or after this time.
+    async def _run_loop(
+        self,
+        *,
+        name: str,
+        compute_next: Callable[[], Time],
+        set_time_attr: Callable[[Time], None],
+        condition: asyncio.Condition,
+        ready_event: asyncio.Event,
+    ) -> None:
         """
-        return get_crossing_time(self.sun_altitude, going_up=False, search_from=after)
+        Background loop to schedule and notify a diurnal event.
 
-    async def _run_twilight_loop(self) -> None:
-        """Notifies twilight_condition whenever it's noon.
+        This coroutine repeatedly computes the next occurrence of an event
```

Comment:
Again, use an active voice so `This coroutine repeatedly computes` -> `Compute`.